### ISSUE TITLE
Custom kernel for 2d convolution

### ISSUE THREAD

**Issue by jmmcd** [pytorch/pytorch#8027]
I want to create a 2d convolution with a kernel like this:

```
000010000
000010000
100010001
000010000
000010000
```

The horizontal aspect is like dilation, I guess, but the vertical part is different. I see that dilation is available as a parameter in conv2d, but it has to be a scalar or single-element tuple (not one element per dimension), so I don't think it can do what I want here. More generally, I would like to be able to specify a kernel as an arbitrary matrix.

(For reference, I added the same question on Stackoverflow, but I now guess this is really a new feature request, not a how to? request: https://stackoverflow.com/questions/50635736/custom-convolution-kernel-and-toroidal-convolution-in-pytorch)

**Comment by apaszke** [comment]
cc @lantiga who figured out an elegant solution to work with custom Conv kernels IIRC

**Comment by fmassa** [comment]
@jmmcd note that an easy (but not the most efficient) way of doing this is by simply doing a masking on the convolution kernel, using the functional interface.

Here is an example:

```python
class MyConv2d(nn.Conv2d):
    def forward(self, x):
        mask = self.mask # a matrix with the masking pattern
        mask = mask[None, None] # mask is 2d, unsqueeze for broadcast
        return F.conv2d(x, self.weight * mask, ...)
```

**Comment by SsnL** [comment]
You can specify the kernel as an arbitrary matrix using the functional interface @fmassa mentioned above.

**Comment by lantiga** [comment]
Hey @jmmcd, @apaszke was referring to this https://gist.github.com/lantiga/a78581e6c6c0ad1534065950e204ce9d
Let me know if you have questions.

**Comment by soumith** [comment]
the above suggested workarounds are the way to go.
